File: SourceAFIS/Simple/Person.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;
using System.Xml.Serialization;

namespace SourceAFIS.Simple
{
    /// <summary>
    /// Collection of fingerprints belonging to one person.
    /// </summary>
    /// <remarks>
    /// <para>
    /// This class is primarily a way to group multiple fingerprints belonging to one person.
    /// This is very convenient feature when there are multiple fingerprints per person, because
    /// it is possible to match two persons directly instead of iterating over their fingerprints.
    /// </para>
    /// <para>
    /// Id field is provided as simple means to bind Person objects to application-specific
    /// information. If you need more flexibility, inherit from Person class and add
    /// application-specific fields as necessary.
    /// </para>
    /// <para>
    /// This class is designed to be easy to serialize in order to be stored in binary format (BLOB)
    /// in application database, binary or XML files, or sent over network. You can either serialize
    /// the whole Person or serialize individual fingerprints.
    /// </para>
    /// </remarks>
    [Serializable]
    public class Person : IList<Fingerprint>, ICloneable
    {
        /// <summary>
        /// Application-assigned ID for the person.
        /// </summary>
        [XmlAttribute]
        public int Id;

        List<Fingerprint> InnerList = new List<Fingerprint>();

        /// <summary>
        /// Get/set all fingerprints belonging to this person as a collection.
        /// </summary>
        public Fingerprint[] AllFingerprints
        {
            get { return InnerList.ToArray(); }
            set { InnerList = new List<Fingerprint>(value); }
        }

        /// <summary>
        /// Creates empty Person object.
        /// </summary>
        public Person() { }

        /// <summary>
        /// Number of fingerprints belonging to the person.
        /// </summary>
        public int Count { get { return InnerList.Count; } }
        /// <summary>
        /// Add fingerprint to person's fingerprint collection.
        /// </summary>
        /// <param name="fp"></param>
        public void Add(Fingerprint fp) { CheckNull(fp); InnerList.Add(fp); }
        /// <summary>
        /// Remove all fingerprints from person's fingerprint collection.
        /// </summary>
        public void Clear() { InnerList.Clear(); }
        /// <summary>
        /// Remove fingerprint from person's fingerprint collection.
        /// </summary>
        /// <param name="fp"></param>
        /// <returns></returns>
        public bool Remove(Fingerprint fp) { return InnerList.Remove(fp); }
        /// <summary>
        /// Access person's fingerprint by index.
        /// </summary>
        /// <param name="index"></param>
        /// <returns></returns>
        public Fingerprint this[int index]
        {
            get { return InnerList[index]; }
            set { CheckNull(value); InnerList[index] = value; }
        }
        /// <summary>
        /// Add fingerprint to person's fingerprint collection at specific index.
        /// </summary>
        /// <param name="index"></param>
        /// <param name="fp"></param>
        public void Insert(int index, Fingerprint fp) { CheckNull(fp); InnerList.Insert(index, fp); }
        /// <summary>
        /// Remove fingerprint from person's fingerprint collection at specified index.
        /// </summary>
        /// <param name="index"></param>
        public void RemoveAt(int index) { InnerList.RemoveAt(index); }
        /// <summary>
        /// Create deep copy of the Person.
        /// </summary>
        /// <returns></returns>
        public Person Clone()
        {
            Person clone = new Person();
            foreach (Fingerprint fp in InnerList)
                clone.Add(fp.Clone());
            return clone;
        }

        IEnumerator<Fingerprint> IEnumerable<Fingerprint>.GetEnumerator() { return InnerList.GetEnumerator(); }
        IEnumerator IEnumerable.GetEnumerator() { return InnerList.GetEnumerator(); }
        bool ICollection<Fingerprint>.IsReadOnly { get { return false; } }
        bool ICollection<Fingerprint>.Contains(Fingerprint fp) { return InnerList.Contains(fp); }
        void ICollection<Fingerprint>.CopyTo(Fingerprint[] array, int index) { InnerList.CopyTo(array, index); }
        int IList<Fingerprint>.IndexOf(Fingerprint fp) { return InnerList.IndexOf(fp); }
        object ICloneable.Clone() { return Clone(); }

        void CheckNull(Fingerprint fp)
        {
            if (fp == null)
                throw new ArgumentNullException();
        }
    }
}

File: java/sourceafis/src/sourceafis/simple/Person.java
package sourceafis.simple;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

/// <summary>
/// Collection of <see cref="Fingerprint"/>s belonging to one person.
/// </summary>
/// <remarks>
/// <para>
/// This class is primarily a way to group multiple <see cref="Fingerprint"/>s belonging to one person.
/// This is very convenient feature when there are multiple fingerprints per person, because
/// it is possible to match two <see cref="Person"/>s directly instead of iterating over their <see cref="Fingerprint"/>s.
/// </para>
/// <para>
/// <see cref="Id"/> property is provided as simple means to bind <see cref="Person"/> objects to application-specific
/// information. If you need more flexibility, inherit from <see cref="Person"/> class and add
/// application-specific fields as necessary.
/// </para>
/// <para>
/// This class is designed to be easy to serialize in order to be stored in binary format (BLOB)
/// in application database, binary or XML files, or sent over network. You can either serialize
/// the whole <see cref="Person"/> or serialize individual <see cref="Fingerprint"/>s.
/// </para>
/// </remarks>
/// <seealso cref="Fingerprint"/>
 
@SuppressWarnings("serial")
public class Person implements Cloneable,Serializable
{
    /// <summary>
    /// Application-assigned ID for the <see cref="Person"/>.
    /// </summary>
    /// <remarks>
    /// <para>
    /// SourceAFIS doesn't use this property. It is provided for applications as an easy means
    /// to link <see cref="Person"/> objects back to application-specific data. Applications can store any
    /// integer ID in this field, for example database table key or an array index.
    /// </para>
    /// <para>
    /// Applications that need to attach more detailed information to the person should
    /// inherit from <see cref="Person"/> class and add fields as necessary.
    /// </para>
    /// </remarks>
    //[XmlAttribute]
    public int Id; // { get; set; }

    List<Fingerprint> FingerprintList = new ArrayList<Fingerprint>();

    /// <summary>
    /// List of <see cref="Fingerprint"/>s belonging to the <see cref="Person"/>.
    /// </summary>
    /// <remarks>
    /// This collection is initially empty. Add <see cref="Fingerprint"/> objects
    /// here. You can also assign the whole collection.
    /// </remarks>
    public List<Fingerprint> fingerprints(){
          return FingerprintList; 
     }
    

    /// <summary>
    /// Creates empty <see cref="Person"/> object.
    /// </summary>
    public Person()
    {
    }

    /// <summary>
    /// Creates new <see cref="Person"/> object and initializes it with
    /// a list of <see cref="Fingerprint"/>s.
    /// </summary>
    /// <param name="fingerprints"><see cref="Fingerprint"/> objects to add to the new <see cref="Person"/>.</param>
    public Person(Fingerprint... fingerprints)
    {
    	for(Fingerprint fp:fingerprints){
    		FingerprintList.add(fp);
    	}
        //this.Fingerprints =  fingerprints.ToList();
    }

    /// <summary>
    /// Create deep copy of the <see cref="Person"/>.
    /// </summary>
    /// <returns>Deep copy of the <see cref="Person"/>.</returns>
    /// <remarks>
    /// This method also clones all <see cref="Fingerprint"/> objects contained
    /// in this <see cref="Person"/>.
    /// </remarks>
   
    /*public Person Clone()
    {
        Person clone = new Person();
        clone.Id = Id;
        foreach (Fingerprint fp in FingerprintList)
            clone.Fingerprints.Add(fp.Clone());
        return clone;
    }
 */
    // ICloneable.Clone() { return Clone(); }

    public Person clone() {
    	try {
    		Person copy = (Person)super.clone();
    		copy.FingerprintList = new ArrayList<Fingerprint>(FingerprintList);
    		return copy;
    	} catch (CloneNotSupportedException e) {
    		throw new RuntimeException(e);
    	}
    }

    void CheckForNulls()
    {
        for (Fingerprint fp : FingerprintList)
            if (fp == null)
                throw new RuntimeException("Person contains null Fingerprint references.");
    }
}